### PR TITLE
Move Quotation validation call from TaxJar to AWC

### PR DESCRIPTION
**Pull request by Alchez**
Connected to https://github.com/DigiThinkIT/awesome_cart/pull/233

---

If a wrong address is entered, TaxJar (or PyCountry) throws an error before a Shipment API call is made.

---

image

### CODE REVIEW

**Comment by vjFaLk** [comment]
Change the error message to something more relevant

File: erpnext_taxjar/hooks.py [hunk 1]
```diff
@@ -16,19 +16,19 @@
...
 	},
-    "Sales Invoice" : {
+	"Sales Invoice" : {
 		"validate" : "erpnext_taxjar.api.set_sales_tax",
 		"on_submit" : "erpnext_taxjar.api.create_transaction",
 		"on_cancel" : "erpnext_taxjar.api.delete_transaction"
 	}
 }
 
+taxjar_api = [
```
Comment (vjFaLk) [inline erpnext_taxjar/hooks.py]: This goes in AWC not here.

Comment (vjFaLk) [inline erpnext_taxjar/hooks.py]: Rename this to "tax_api"

File: erpnext_taxjar/hooks.py [hunk 2]
```diff
@@ -16,19 +16,19 @@
 
 
 doc_events = {
-	"Quotation" : {
-		"validate": "erpnext_taxjar.api.set_sales_tax"
-	},
```
Comment (DeveloperMetal) [inline erpnext_taxjar/hooks.py]: Question... won't this remove taxes from quotations on the backend if they start one manually?

Do we know if this is a problem?

File: erpnext_taxjar/hooks.py [hunk 3]
```diff
@@ -16,19 +16,19 @@
...
+	"Sales Invoice" : {
 		"validate" : "erpnext_taxjar.api.set_sales_tax",
 		"on_submit" : "erpnext_taxjar.api.create_transaction",
 		"on_cancel" : "erpnext_taxjar.api.delete_transaction"
 	}
 }
 
+tax_api = [
+	"erpnext_taxjar.api.set_sales_tax"
+]
```
Comment (DeveloperMetal) [inline erpnext_taxjar/hooks.py]: See my comment on awc about renaming the hook

**Comment by Alchez** [comment]
I've separated the error handling for the addresses outside setting and getting sales tax, so that they can be called independently.

So AWC will call `validate_address`, which does a `validate_state` call (for PyCountry error handling) and a `validate_tax_request` call (for TaxJar error handling), if any are caught.

**Comment by Alchez** [comment]
Keeping the Quotation hook for now, since it will not apply Sales Tax for backend orders, but we'll need some more discussion on how to reduce TaxJar calls since a Quotation is saved whenever you change anything in the cart.

**Comment by DeveloperMetal** [comment]
So while testing I got this:

```
frappe-web.min.js:988 Traceback (most recent call last):
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/app.py", line 62, in application
    response = frappe.handler.handle()
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/handler.py", line 22, in handle
    data = execute_cmd(cmd)
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/handler.py", line 53, in execute_cmd
    return frappe.call(method, **frappe.form_dict)
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/__init__.py", line 939, in call
    return fn(*args, **newargs)
  File "/home/vagrant/frappe-bench/apps/awesome_cart/awesome_cart/awc.py", line 1090, in cart
    quotation.save()
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/model/document.py", line 256, in save
    return self._save(*args, **kwargs)
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/model/document.py", line 290, in _save
    self.run_before_save_methods()
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/model/document.py", line 859, in run_before_save_methods
    self.run_method("validate")
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/model/document.py", line 755, in run_method
    out = Document.hook(fn)(self, *args, **kwargs)
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/model/document.py", line 1024, in composer
    return composed(self, method, *args, **kwargs)
  File "/home/vagrant/frappe-bench/apps/frappe/frappe/model/document.py", line 1009, in runner
    add_to_return_value(self, f(self, method, *args, **kwargs))
  File "/home/vagrant/frappe-bench/apps/erpnext_taxjar/erpnext_taxjar/api.py", line 143, in set_sales_tax
    tax_dict = get_tax_data(doc)
  File "/home/vagrant/frappe-bench/apps/erpnext_taxjar/erpnext_taxjar/api.py", line 90, in get_tax_data
    shipping_state = validate_state(shipping_address)
  File "/home/vagrant/frappe-bench/apps/erpnext_taxjar/erpnext_taxjar/api.py", line 198, in validate_state
    states = pycountry.subdivisions.get(country_code=country_code)
  File "/home/vagrant/frappe-bench/env/local/lib/python2.7/site-packages/pycountry/__init__.py", line 108, in get
    countries.get(alpha_2=kw['country_code'])
  File "/home/vagrant/frappe-bench/env/local/lib/python2.7/site-packages/pycountry/db.py", line 43, in load_if_needed
    return f(self, *args, **kw)
  File "/home/vagrant/frappe-bench/env/local/lib/python2.7/site-packages/pycountry/db.py", line 101, in get
    return self.indices[field][value]
KeyError: u'us'
```

Tested with all repos on develop branch except for awesome_cart and this repo using your pull requests.
Am I missing something?

**Comment by Alchez** [comment]
Whoops, needed to uppercase the country code before passing it, since Frappe stores them in lowercase.

Also, I forgot to mention you will need [Valmik's fix](https://github.com/DigiThinkIT/shipment_management/pull/76) for retrieving the shipping rates.

**Comment by DeveloperMetal** [comment]
I found Two issues with the implementation:

1) If you edit an existing address and set the state to an invalid state say Florida -> Foorida. Validation kicks in and lets you know the address is wrong but will allow the address to say wrong and lets you continue checking out.
2) After the above was allowed, adding creditcard information to checkout with an invalid address triggered validation again and correctly stopped checkout... BUT... the card was still charged and I got an authorize.net receipt(as the owner of the test authnet account) of the transaction. But the cart only displayed the validation error and made it appear as if there was no transaction.

The first issue is probably easier to fix, we just can't allow the address to stay with an invalid value if the validation broke.

The second one is partly caused by the above one, but validation needs to stop the transaction process too.

Also, the error box on checkout has a { white-space: pre-wrap } css attribute set. Please remove tabs and consider formatting with css instead. It looks horrible on smaller screens and mobile layout

**Comment by DeveloperMetal** [comment]
Here is a snapshot of the checkout on a smaller screen:

address-validation-error

**Comment by DeveloperMetal** [comment]
Rechecked my setup and the previous errors aren't there. Sorry about that.  :D

However, there there is a validation issue if you enter all lowercase states:

address-validation-error-2

Otherwise, if you enter all caps "FL" or the full state "Florida", "florida" or "FLORIDA" there no issues.

**Comment by Alchez** [comment]
Whoops, added the validation for lower-case states.

**Review by DeveloperMetal — DISMISSED** [review]

**Review by vjFaLk — DISMISSED** [review]

**Review by DeveloperMetal — COMMENTED** [review]
Tested the changes but now I get no errors if I enter an invalid(ex: "WAT") state.

It treats the address as if there was no taxes to charge.

**Review by vjFaLk — APPROVED** [review]

**Review by DeveloperMetal — APPROVED** [review]

**Pull request merged**